client/my-sites/marketing/buttons/use-sharing-buttons-query.js

```diff
@@ -0,0 +1,50 @@
+import { useCallback } from 'react';
+import { useMutation, useQuery, useQueryClient } from 'react-query';
+import wpcom from 'calypso/lib/wp';
+
+export function useSharingButtonsQuery( siteId ) {
+	return useQuery(
+		[ 'sharing-buttons', siteId ],
+		() => wpcom.req.get( `/sites/${ siteId }/sharing-buttons` ),
+
+		{
+			enabled: !! siteId,
+			select( data ) {
+				return data.sharing_buttons;
+			},
+		}
+	);
+}
+
+export function useSaveSharingButtonsMutation( siteId ) {
+	const queryClient = useQueryClient();
+	const mutation = useMutation(
+		( { buttons } ) =>
+			wpcom.req.post( `/sites/${ siteId }/sharing-buttons`, { sharing_buttons: buttons } ),
+		{
+			async onMutate( { buttons } ) {
+				// optimistic update
```

Comment:
I know the Redux version did these optimistic updates, too, but are they really necessary? The `<SharingButtons>` component kind of does them already by storing the interim dirty `buttons` array in `this.state.buttonsPendingSave` and rendering either the dirty version or the saved version:
```js
const updatedButtons = this.state.buttonsPendingSave || buttons;
```

Reply:
I didn't include the optimistic update in my original migration but found that there's a jump between old and new state after saving (see capture below). We set `buttonsPendingSave` to `null` at a point where the invalidated query has not been refetched yet.

https://user-images.githubusercontent.com/9202899/135582101-443f1a43-4c56-4f45-ae8e-f5d78654e216.mp4

Reply:
That's a good point, I didn't realize `buttonsPendingSave` would get reset before the query is refetched.

But: I was testing the feature now and noticed that the `POST /sites/:site/sharing-buttons` returns the updated buttons array as part of the response: it has shape `{ success: true, updated: [ ...buttons ] }`. That means we don't need to invalidate the query results and trigger refetch. We can use the `response.updated` array to update the query results immediately when the mutation finishes. That would simplify things a lot.

Reply:
Added https://github.com/Automattic/wp-calypso/pull/56579/commits/064842e5dbdc71c7ce59b7d7c9c4bc201618c3ab. Let me know what you think.